datayoga/blocks/expression.py

```diff
@@ -32,12 +32,58 @@ def search(self, data: Any) -> Any:
         """
         pass
 
+    def test(self, data: Any) -> bool:
+        """Test a where clause for an SQL statement
+
+        Args:
+            data (Any): Data
+
+        Returns:
+            boolean: True if matches, else False
+        """
+        pass
+
 
 class SQLExpression(Expression):
     def compile(self, expression: str):
         self.conn = sqlite3.connect(":memory")
         self.expression = expression
 
+    def filter(self, data: List[Any]) -> List[Any]:
+        """Test a where clause for an SQL statement
+
+        Args:
+            data (Any): Data
+
+        Returns:
+            boolean: True if matches, else False
+        """
+        # use a CTE to create the in memory data structure
+        cte_clause = self._get_cte(data)
+
+        column_names = data[0].keys()
+        # fetch the CTE and bind the variables
+        data_values = [row.get(colname) for row in data for colname in column_names]
+        self.conn.row_factory = sqlite3.Row
+
+        cursor = self.conn.execute(
+            f"{cte_clause} select * from data where {self.expression}", data_values
+        )
+        return [dict(row) for row in cursor.fetchall()]
+
+    def _get_cte(self, data: List[Any]) -> str:
+        # builds a CTE expression for fetching in memory data
+
+        column_names = data[0].keys()
+        columns_clause = ','.join(f"`{col}`" for col in column_names)
+
+        # values in the form of (?,?), (?,?)
+        values_clause_row = f"({','.join('?'*len(column_names))})"
+        values_clause = ','.join([values_clause_row]*len(data))
+
+        # use a CTE to create the in memory data structure
+        return f"with data({columns_clause}) as (values {values_clause})"
+
     def search(self, data: Any) -> Any:
```

Comment:
should be List[Dict[str, Any]] (both places)

Reply:
done